ui/src/app/views/workspace-share/component.tsx

```diff
@@ -0,0 +1,409 @@
+    height: '4.5%',
+    color: '#888',
+    fontFamily: 'Montserrat',
+    lineHeight: '1rem',
+    whiteSpace: 'nowrap'
+  },
+
+  collaboratorIcon: {
+    margin: '0 0 0 4rem',
+    color: '#2691D0',
+    cursor: 'pointer'
+  },
+
+  sharingBody: {
+    maxHeight: '70vh',
+    overflowY: 'hidden',
+    overflowX: 'visible',
+    padding: '0 .125rem',
+    display: 'flex',
+    flexDirection: 'column'
+  }
+
+});
+
+export interface WorkspaceShareState {
+  autocompleteLoading: boolean;
+  autocompleteUsers: User[];
+  userNotFound: string;
+  workspaceShareError: boolean;
+  usersLoading: boolean;
+  userRolesList: UserRole[];
+  workspaceFound: boolean;
+  workspaceUpdateConflictError: boolean;
+  workspace: Workspace;
+  searchTerm: string;
+}
+
+export interface WorkspaceShareProps {
+  workspace: Workspace;
+  accessLevel: WorkspaceAccessLevel;
+  userEmail: string;
+  closeFunction: Function;
+  sharing: boolean;
+}
+
+export class WorkspaceShare extends React.Component<WorkspaceShareProps, WorkspaceShareState> {
+
+  constructor(props: WorkspaceShareProps) {
+    super(props);
+    this.state = {
+      autocompleteLoading: false,
+      autocompleteUsers: [],
+      userNotFound: '',
+      workspaceShareError: false,
+      usersLoading: false,
+      userRolesList: this.props.workspace.userRoles,
+      workspaceFound: (this.props.workspace !== null),
+      workspaceUpdateConflictError: false,
+      workspace: this.props.workspace,
+      searchTerm: '',
+    };
+    this.searchTermChangedEvent = fp.debounce(300, this.userSearch);
+  }
+
+  save(): void {
+    if (this.state.usersLoading) {
+      return;
+    }
+    this.setState({usersLoading: true, workspaceShareError: false});
+    workspacesApi().shareWorkspace(this.state.workspace.namespace,
+      this.state.workspace.id,
+      {workspaceEtag: this.state.workspace.etag, items: this.state.workspace.userRoles})
+      .then((resp: ShareWorkspaceResponse) => {
+        const updatedWorkspace = this.state.workspace;
+        fp.set('etag', resp.workspaceEtag, updatedWorkspace);
+        fp.set('userRoles', resp.items, updatedWorkspace);
+        this.setState({usersLoading: false, userNotFound: '',
+          searchTerm: '', workspace: updatedWorkspace});
+        this.props.closeFunction();
+      }).catch(error => {
+        if (error.status === 400) {
+          this.setState({userNotFound: error});
+        } else if (error.status === 409) {
+          this.setState({workspaceUpdateConflictError: true});
+        } else {
+          this.setState({workspaceShareError: true});
+        }
+        this.setState({usersLoading: false});
+      });
+  }
+
+  removeCollaborator(user: UserRole): void {
+    if (!this.state.usersLoading) {
+      const position = this.state.userRolesList.findIndex((userRole) => {
+        return user.email === userRole.email;
+      });
+      this.setState(({userRolesList}) => (
+        {userRolesList: fp.pullAt(position, userRolesList)}));
```

Comment:
`fp.remove(({email}) => user.email === email, userRolesList)` is probably easier than the two step thing